config_dump.txt
{
  "version": 1,
  "controllers": {
    "con1": {
      "analog_outputs": {
        "1": {
          "offset": 0.0
        },
        "2": {
          "offset": 0.0
        },
        "3": {
          "offset": 0.0
        },
        "4": {
          "offset": 0.0
        },
        "5": {
          "offset": 0.0
        },
        "6": {
          "offset": 0.0
        },
        "7": {
          "offset": 0.0
        },
        "8": {
          "offset": 0.0
        },
        "9": {
          "offset": 0.0
        },
        "10": {
          "offset": 0.0
        }
      },
      "digital_outputs": {
        "1": {},
        "2": {},
        "3": {},
        "4": {},
        "5": {},
        "6": {},
        "7": {},
        "8": {},
        "9": {},
        "10": {}
      },
      "analog_inputs": {
        "1": {
          "offset": 0.0129,
          "gain_db": 0
        },
        "2": {
          "offset": 0.0,
          "gain_db": 0
        }
      }
    }
  },
  "elements": {
    "generic_output_elem_ch1_ch1_1": {
      "singleInput": {
        "port": [
          "con1",
          1
        ]
      },
      "intermediate_frequency": 10000000,
      "operations": {
        "readout": "meas_pulse_out_ch1_in_ch1_1"
      },
      "time_of_flight": 24,
      "smearing": 0,
      "outputs": {
        "out1": [
          "con1",
          1
        ]
      }
    },
    "generic_ai_elem_ch3_1": {
      "singleInput": {
        "port": [
          "con1",
          3
        ]
      },
      "intermediate_frequency": 50000000,
      "operations": {
        "const": "square_pulse_ch3_1"
      }
    }
  },
  "pulses": {
    "meas_pulse_out_ch1_in_ch1_1": {
      "operation": "measurement",
      "length": 500,
      "waveforms": {
        "single": "square_measure_wf_ch1_1"
      },
      "integration_weights": {
        "cos": "cos",
        "sin": "sin",
        "minus_sin": "minus_sin"
      },
      "digital_marker": "ON"
    },
    "square_pulse_ch3_1": {
      "operation": "control",
      "length": 300,
      "waveforms": {
        "single": "square_wf_ch3_1"
      }
    }
  },
  "waveforms": {
    "square_wf": {
      "type": "constant",
      "sample": 0.1
    },
    "gaussian_wf": {
      "type": "arbitrary",
      "samples": [
        0.0,
        0.001827960579371176,
        0.003849392974211451,
        0.006078226182377813,
        0.008528495159196965,
        0.011214205375200008,
        0.014149181085316952,
        0.01734689773231606,
        0.02082029925124469,
        0.02458160141031431,
        0.028642082711234234,
        0.03301186477142755,
        0.03769968451374146,
        0.042712660887083295,
        0.04805605922391891,
        0.05373305669716996,
        0.05974451265873593,
        0.06608874791351863,
        0.07276133719548414,
        0.07975491925551224,
        0.08705902903496318,
        0.09465995637564234,
        0.10254063559932486,
        0.11068057007324379,
        0.11905579555914715,
        0.12763888572231175,
        0.13639900265549773,
        0.14530199465621918,
        0.15431054279166792,
        0.16338435700470996,
        0.1724804216697736,
        0.18155328961481923,
        0.19055542270272005,
        0.19943757613187663,
        0.20814922269264116,
        0.21663901232485316,
        0.2248552614844414,
        0.23274646606521499,
        0.24026183095625636,
        0.24735180876473745,
        0.25396863981532036,
        0.2600668852646343,
        0.2656039450534739,
        0.2705405524675425,
        0.27484123729302445,
        0.27847474993514093,
        0.28141443941104427,
        0.28363857882368865,
        0.2851306327574691,
        0.28587946199252734,
        0.28587946199252734,
        0.2851306327574691,
        0.28363857882368865,
        0.28141443941104427,
        0.27847474993514093,
        0.27484123729302445,
        0.2705405524675425,
        0.2656039450534739,
        0.2600668852646343,
        0.25396863981532036,
        0.24735180876473745,
        0.24026183095625636,
        0.23274646606521499,
        0.2248552614844414,
        0.21663901232485316,
        0.20814922269264116,
        0.19943757613187663,
        0.19055542270272005,
        0.18155328961481923,
        0.1724804216697736,
        0.16338435700470996,
        0.15431054279166792,
        0.14530199465621918,
        0.13639900265549773,
        0.12763888572231175,
        0.11905579555914715,
        0.11068057007324379,
        0.10254063559932486,
        0.09465995637564234,
        0.08705902903496318,
        0.07975491925551224,
        0.07276133719548414,
        0.06608874791351863,
        0.05974451265873593,
        0.05373305669716996,
        0.04805605922391891,
        0.042712660887083295,
        0.03769968451374146,
        0.03301186477142755,
        0.028642082711234234,
        0.02458160141031431,
        0.02082029925124469,
        0.01734689773231606,
        0.014149181085316952,
        0.011214205375200008,
        0.008528495159196965,
        0.006078226182377813,
        0.003849392974211451,
        0.001827960579371176,
        0.0
      ]
    },
    "square_measure_wf_ch1_1": {
      "type": "constant",
      "sample": 0.3
    },
    "square_wf_ch3_1": {
      "type": "constant",
      "sample": 0.4
    }
  },
  "digital_waveforms": {
    "ON": {
      "samples": [
        [
          1,
          0
        ]
      ]
    }
  },
  "integration_weights": {
    "cos": {
      "cosine": [
        [
          4.0,
          500
        ]
      ],
      "sine": [
        [
          0.0,
          500
        ]
      ]
    },
    "sin": {
      "cosine": [
        [
          0.0,
          500
        ]
      ],
      "sine": [
        [
          4.0,
          500
        ]
      ]
    },
    "minus_sin": {
      "cosine": [
        [
          0.0,
          20
        ]
      ],
      "sine": [
        [
          -4.0,
          20
        ]
      ]
    }
  }
}

pylabnet/qua_practice/configuration.py
from pathlib import Path
import numpy as np
from qualang_tools.config.waveform_tools import drag_gaussian_pulse_waveforms
from qualang_tools.units import unit

# These packages are imported here so that we don't have to import them in all the other files
from qualang_tools.plot import interrupt_on_close
from qualang_tools.results import progress_counter, fetching_tool
import plotly.io as pio

pio.renderers.default = "browser"

#######################
# AUXILIARY FUNCTIONS #
#######################
u = unit(coerce_to_integer=True)
# Used to correct for IQ mixer imbalances


######################
# Network parameters #
######################
# IP address of the Quantum Orchestration Platform
qop_ip = "192.168.88.252"
cluster_name = "Cluster_1"


#############
# Save Path #
#############
# Path to save data
save_dir = Path(__file__).parent.resolve() / "Data"
save_dir.mkdir(exist_ok=True)

default_additional_files = {
    Path(__file__).name: Path(__file__).name,
    "optimal_weights.npz": "optimal_weights.npz",
}

#####################
# OPX configuration #
#####################

# Frequencies
AOM_IF = 10 * u.MHz

# Gaussian pulse parameters
gauss_amp = 0.3  # The gaussian is used when calibrating pi and pi_half pulses
gauss_len = 100  # The gaussian is used when calibrating pi and pi_half pulses
gauss_wf, gauss_der_wf = drag_gaussian_pulse_waveforms(
    gauss_amp, gauss_len, gauss_len / 5, alpha=0, delta=1, anharmonicity=0, detuning=0, subtracted=True
)

# Square pulse parameters
square_len = 100 # in ns
square_amp = 0.5 # in Volts

# Trigger pulse parameters
trigger_len = 320  # in ns

config = {
    "version": 1,
    "controllers": {
        "con1": {
            "analog_outputs": {
                1: {"offset": +0.0},
                2: {"offset": +0.0},
            },
            "digital_outputs": {
                1: {},
                10: {},
            },
            "analog_inputs": {
                1: {"offset": 0.0, "gain_db": 0},
                2: {"offset": 0.0, "gain_db": 0},
            },
        },
    },
    "elements": {
        "AOM": {
            "singleInput": {"port": ("con1", 1)},
            "intermediate_frequency": AOM_IF,
            "operations": {
                "const": "square_pulse",
                "gauss": "gaussian_pulse",
            }
        },
        "AOM_2": {
            "singleInput": {"port": ("con1", 2)},
            "intermediate_frequency": AOM_IF,
            "operations": {
                "const": "square_pulse",
                "gauss": "gaussian_pulse",
            }
        },
        # "trigger": {
        #     "digitalInputs": {
        #         "activate": {
        #             "port": ("con1", 1),
        #             "delay": 0,
        #             "buffer": 0,
        #         },
        #         #"activate2": {
        #         #    "port": ("con1", 10),
        #         #    "delay": 0,
        #         #    "buffer": 0,
        #         #},
        #     },
        #     "operations": {
        #         "ON": "ON_pulse",
        #     },
        # },
        "generic_output_elem": {
            "singleInput": {"port": ("con1", 1)},
            "intermediate_frequency": AOM_IF,
            "operations": {
                "readout": "meas_pulse_in",
            },
            'time_of_flight': 180,
            'smearing': 0,
            'outputs': {
                'out1': ('con1', 1)
            }
        }
    },
    "pulses": {
        'meas_pulse_in': {
            'operation': 'measurement',
            'length': 250,
            'waveforms': {
                'single': 'square_wf'
            },
            'integration_weights': { #doesn't get used if we're not demodulating
                'cos': 'cos',
                'sin': 'sin',
                'minus_sin': 'minus_sin',
            },
            'digital_marker': 'ON'  #why is this needed? Is it needed?
        },
        "ON_pulse": {
            "operation": "control",
            "length": trigger_len,
            "digital_marker": "ON",
        },
        "gaussian_pulse": {
            "operation": "control",
            "length": gauss_len,  # in ns
            "waveforms": {"single": "gaussian_wf"},
        },
        "square_pulse": {"operation": "control", "length": square_len, "waveforms": {"single": "square_wf"}},  # in ns
    },
    "waveforms": {
        "square_wf": {"type": "constant", "sample": square_amp},
        "gaussian_wf": {"type": "arbitrary", "samples": gauss_wf},
    },
    "digital_waveforms": {
        "ON": {"samples": [(1, 0)]},
    },
    'integration_weights': {
        'cos': {
            'cosine': [(4.0, 20)],
            'sine': [(0.0, 20)]
        },
        'sin': {
            'cosine': [(0.0, 20)],
            'sine': [(4.0, 20)]
        },
        'minus_sin': {
            'cosine': [(0.0, 20)],
            'sine': [(-4.0, 20)]
        },
    },
}
